### ISSUE TITLE
[Question] How to define a time-varying external force

### ISSUE THREAD

**Issue by JinxrDK** [openmm/openmm#4305]
Dear developers, now I want to define a sinusoidal electric field, I followed the comments in issue #3066. The static electric field works well, but what should I do to import **Simulation Time** to the force. My current code is as followed.

`
# Convert unit1 (V/nm) to unit2 (kJ/(mol*nm*e)) 
def Voltage_per_nm_to_kJ_mol(EFintensity):
    EF_converted = EFintensity * volt / nanometers / item
    return EF_converted.in_units_of(kilojoules_per_mole/(nanometers * elementary_charge))
# Convert Frequency (1/s) to Angular frequency (2*pi/s)
def Frequency_add_unit(Frequency):
    mu = Frequency * 2 * math.pi / second
    return mu

def ExternalElectricField(system, inputs):
    # Calculate electric field force
    ExternalEF = CustomExternalForce('-q*Ex*cos(miux*t)*x-q*Ey*cos(miuy*t)*y-q*Ez*cos(miuz*t)*z') 
    # Convert the Unit
    ef_x_converted = Voltage_per_nm_to_kJ_mol(inputs.ef_x) 
    ef_y_converted = Voltage_per_nm_to_kJ_mol(inputs.ef_y)
    ef_z_converted = Voltage_per_nm_to_kJ_mol(inputs.ef_z)

    ExternalEF.addPerParticleParameter("q")
    ExternalEF.addGlobalParameter("Ex",ef_x_converted)
    ExternalEF.addGlobalParameter("Ey",ef_y_converted)
    ExternalEF.addGlobalParameter("Ez",ef_z_converted)

    ExternalEF.addGlobalParameter("t") #How to import Simulation Time?
    ExternalEF.addGlobalParameter("miux",Frequency_add_unit(inputs.frequency_x))
    ExternalEF.addGlobalParameter("miuy",Frequency_add_unit(inputs.frequency_y))
    ExternalEF.addGlobalParameter("miuz",Frequency_add_unit(inputs.frequency_z))
    
    # Obtain atomic charge of every atoms and apply electric field force on all atoms
    nonbonded = [f for f in system.getForces() if isinstance(f, NonbondedForce)][0]
    for i in range(system.getNumParticles()):
        charge, sigma, epsilon = nonbonded.getParticleParameters(i)
        ExternalEF.addParticle(i, [charge])
    # Add the customized force to system
    system.addForce(ExternalEF)
    return system
`

**Comment by peastman** [comment]
You can call `setParameter()` on the context to modify the value of a global parameter.  You can do something like this.

```python
for i in range(steps):
    simulation.context.setParameter('t', simulation.context.getTime())
    simulation.step(1)
```

**Comment by peastman** [comment]
This question does come up pretty often.  The solution shown above is easy and fairly efficient, but it might still be worth adding direct support for a time variable in the custom force classes to make it even easier.  I'll mark this as a feature request.

**Comment by JinxrDK** [comment]
Thanks for your reply.

**Comment by peastman** [comment]
Let's keep this open.  I've marked it as a feature request.

**Comment by peastman** [comment]
I implemented this for most of the commonly used custom forces.  But having done it, I'm now having second thoughts about whether it's a good idea.  The problem is that, unless you run in double precision mode, forces are calculated in single precision.  That means the time gets represented in single precision, and calculations on it are done in single precision.

Let's take the example of a sinusoidally varying force:

```python
force = CustomExternalForce('sin(100*t)*x')
```

Here's the force magnitude over the first 25 steps, assuming a 4 fs step size.

![image](https://github.com/openmm/openmm/assets/4379786/8e2cd650-aa39-4fbb-82a9-1e396901e044)

And here's the force magnitude after 100 million steps (400 ns).

![image](https://github.com/openmm/openmm/assets/4379786/d594d8e6-7b6e-4fa7-a6ce-eca31f016907)

Once the time gets too large, it can no longer be represented accurately enough in single precision.

Of course, if you create a parameter for time as suggested above you'll have exactly the same problem.  But at least in that case there's a workaround.  You can define your force as `'f*x'`, then set `f` directly:

```python
simulation.context.setParameter('f', sin(100*simulation.context.getTime()))
```

That will compute it in double precision.

By building in support for it, we're encouraging users to do something that is guaranteed to fail if they just run their simulations long enough.

**Comment by JinxrDK** [comment]
Yes, it's really a question need further consideration. Maybe I think for a long-time simulation, applied time-dependent external force is usually periodic. Thus acquiring the timestep by simulation.context.getStepCount(), and regard the remainder of it divided by period as a new time variable might be better. However, to they who want to use OpenMM as a molecular dynamics black box, the question may be dangerous.

**Comment by peastman** [comment]
We decided against implementing this due to the precision issues discussed above.  There are workarounds for doing the same thing in a more robust way.  For example, if you want a force to vary periodically in time, you can add a global parameter that your python script sets on each step to the current phase within the current period.
